src/recovery/operators.rs

```diff
@@ -179,27 +179,92 @@ where
+                    incoming.swap(&mut tmp_incoming);
+                    // We have to drain the incoming data, but we just
+                    // care about the epoch so drop it.
+                    tmp_incoming.clear();
+                });
+
+                cap = if let Some(cap) = cap.clone() {
+                    let frontier = input_frontiers[0].frontier().iter().min().copied();
+
+                    // Do not delay cap before the write; we will
+                    // delay downstream progress messages longer than
+                    // necessary if so. This would manifest as
+                    // resuming from an epoch that seems "too
+                    // early". Write out the progress at the end of
+                    // each epoch and where the frontier has moved.
+                    let should_write = if let Some(frontier) = frontier {
+                        frontier > *cap.time()
+                    } else {
+                        true
+                    };
+
+                    if should_write {
+                        let write_frontier = if let Some(frontier) = frontier {
+                            frontier
+                        } else {
+                            // There's no way to guarantee this is
+                            // actually the resume epoch on the next
+                            // execution, but mark that this worker is
+                            // ready to resume there on EOF. It's also
+                            // possible that this results in a "too
+                            // small" resume epoch: if for some reason
+                            // this operator isn't activated during
+                            // every epoch, we might miss the
+                            // "largest" epoch and so we'll mark down
+                            // the resume epoch as one too
+                            // small. That's fine, we just might
+                            // resume further back than is optimal.
+                            let resume_epoch = *cap.time() + 1;
+                            resume_epoch
+                        };
+                        let msg = ProgressMsg::Advance(WorkerFrontier(write_frontier));
+                        let kchange = KChange(worker_key.clone(), Change::Upsert(msg));
+                        output_wrapper.activate().session(&cap).give(kchange);
+                    }
+
+                    if let Some(frontier) = frontier {
+                        // We should never delay to something like
+                        // frontier + 1, otherwise chained progress
+                        // operators will "drift" forward and GC will
+                        // happen too early.
+                        Some(cap.delayed(&frontier))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                };
```

Comment:
I'd make use of the Option trait more here (`take`, `map`, `map_or`, `unwrap_or_else`, `flatten`), but maybe it's less readable this way?
Just leaving it here as a possible alternative:
```rust
cap = cap
    .take()
    .map(|cap| {
        let frontier = input_frontiers[0].frontier().iter().min().copied();

        let should_write = frontier.map_or(true, |f| f > *cap.time());

        if should_write {
            let write_frontier = frontier.unwrap_or_else(|| *cap.time() + 1);
            let msg = ProgressMsg::Advance(WorkerFrontier(write_frontier));
            let kchange = KChange(worker_key.clone(), Change::Upsert(msg));
            output_wrapper.activate().session(&cap).give(kchange);
        }
        frontier.map(|f| cap.delayed(&f))
    })
    .flatten();
```

Reply:
Oh great! Thanks. I forget about `take` and `map_or` and that makes it easier.